File: tree.cpp
#include<iostream>
#include<conio.h>
#include<stdlib.h>
using namespace std;
void insertion(int,int );
void delte(int);
int searching(int);
int searching1(int,int);
void display(int);
int tree[40],t=1,s,x,i;
int main()
{
    int ch,y;
    for(i=1; i<40; i++)
        tree[i]=-1;
    while(1)
    {
        cout  <<"\n1.insertion\n2.deletion\n3.DISPLAY\n4.searching\n5.Exit\nEnter Choice:";
        cin >> ch;
        switch(ch)
        {
        case 1:
            cout <<"Enter  element to insertion";
            cin >> ch;
            insertion(1,ch);
            break;
        case 2:
            cout <<"Enter  element to deletion";
            cin >>x;
            y=searching(1);
            if(y!=-1) delte(y);
            else cout<<"No  Element Found ";
            break;
        case 3:
            display(1);
            break;
            display(1);
            cout<<"\n";
            for(i=0; i<=32; i++)
                cout <<i;
            cout <<"\n";
            break;
        case 4:
            cout <<"Enter the Element to searching:";
            cin >> x;
            y=searching(1);
            if(y == -1) cout <<"No such Element Found in Tree";
            else cout <<x << "is in" << y <<"Position";
            break;

        case 5:
            exit(0);
        }
    }
     return 0;
}
void insertion(int s,int ch )
{
    int x;
    if(t==1)
    {
        tree[t++]=ch;
        return;
    }
    x=searching1(s,ch);
    if(tree[x]>ch)
        tree[2*x]=ch;
    else
        tree[2*x+1]=ch;
    t++;
}
void delte(int x)
{
    if( tree[2*x]==-1 && tree[2*x+1]==-1)
        tree[x]=-1;
    else if(tree[2*x]==-1)
    {
        tree[x]=tree[2*x+1];
        tree[2*x+1]=-1;
    }
    else if(tree[2*x+1]==-1)
    {
        tree[x]=tree[2*x];
        tree[2*x]=-1;
    }
    else
    {
        tree[x]=tree[2*x];
        delte(2*x);
    }
    t--;
}
int searching(int s)
{
    if(t==1)
    {
        cout <<"No element in tree";
        return -1;
    }
    if(tree[s]==-1)
        return tree[s];
    if(tree[s]>x)
        searching(2*s);
    else if(tree[s]<x)
        searching(2*s+1);
    else
        return s;
}
void display(int s)
{
    if(t==1)
    {
        cout <<"No element in tree:";
        return;
    }
    for(int i=1; i<40; i++)
        if(tree[i]==-1)
            cout <<" ";
        else cout <<tree[i];
    return ;
}
int searching1(int s,int ch)
{
    if(t==1)
    {
        cout <<"No element in tree";
        return -1;
    }
    if(tree[s]==-1)
        return s/2;
    if(tree[s] > ch)
        searching1(2*s,ch);
    else searching1(2*s+1,ch);
}

File: tree.py
class Tree: 

    def __init__(node, value): 
        node.value = value  
        node.left = None
        node.right = None

    def Inorder( node, Root ): 
        if( Root is None ): 
            return
        node.Inorder(Root.left) 
        print(Root.value,end = ' ') 
        node.Inorder(Root.right) 

    def Insertion(node, value): 
        if node is None: 
            node = Tree(value)
        elif value < node.value:
            if node.left is None:
                node.left = Tree(value)
            else:
               node.left.Insertion(value) 
        else:
            if node.right is None:
                node.right = Tree(value)
            else:
                node.right.Insertion(value)

    def Deletion(node,temp, value):
        if value < node.value:
            temp = node
            node.left.Deletion(temp,value)
        elif(value > node.value):
            temp = node
            node.right.Deletion(temp, value)

        else:
            if (node.left is None and node.right is None):
                if(temp.left == node):
                    temp.left = None
                else:
                    temp.right = None
                node = None

            elif node.right is None :
                if(temp.left == node):
                    temp.left = node.left
                else:
                    temp.right = node.left
                node = None

            elif node.left is None :
                if(temp.left == node):
                    temp.left = node.right
                else:
                    temp.right = node.right
                node = None

            else:
                temp = node.right
                while(temp.left is not None):
                    temp = temp.left 
                node.value = temp.value
                node.right.Deletion(temp,temp.value)


    def searching(self,root ,key):
        if root is None:
            return None
        if root.value == key:
            return root
        elif root.value > key:
            return self.searching(root.left, key)
        else:
            return self.searching(root.right, key)


Root = Tree(5) 
Root.Insertion(3) 
Root.Insertion(2) 
Root.Insertion(1) 
Root.Insertion(8) 
Root.Insertion(4) 


print ("Inorder traversal after insertion: ",end = '')
Root.Inorder(Root) 

Root.Deletion(Root, 4) 
print ('\n 4 is deleted: ',end ='')
Root.Inorder(Root) 

Root.Deletion(Root, 3) 
print ('\n 3 is deleted: ',end ='')
Root.Inorder(Root) 

Root.Deletion(Root, 8) 
print ('\n 8 is deleted: ',end ='')
Root.Inorder(Root)

print()
print("After searching 1")
n = Root.searching(Root,1)
if (n != None):
    print(f"{n.value} Found")
else:
    print("Not present in tree")